constraint-solver/src/quadratic_symbolic_expression.rs

```diff
@@ -208,7 +208,80 @@ impl<T: FieldElement, V: Ord + Clone + Hash + Eq> QuadraticSymbolicExpression<T,
         }
     }
 
-    // TODO implement substitute_by_unknown where substitution is a QSE.
+    /// Substitute a variable by an unknown expression (a QuadraticSymbolicExpression).
+    pub fn substitute_by_unknown(
+        &mut self,
+        variable: &V,
+        substitution: &QuadraticSymbolicExpression<T, V>,
+    ) {
+        if !self.referenced_variables().any(|v| v == variable) {
+            return;
+        }
+
+        match &self.constant {
+            SymbolicExpression::Symbol(sym, _) if sym == variable => {
+                self.constant = SymbolicExpression::from(T::zero());
+                *self += substitution.clone();
+            }
+            _ => {
+                if let Some(mut qse) =
+                    symbolic_expression_to_quadratic_symbolic_expression(&self.constant)
+                {
+                    qse.substitute_by_unknown(variable, substitution);
+                    self.constant = SymbolicExpression::from(T::zero());
+                    *self += qse;
+                }
+            }
+        }
+
+        let mut to_add = QuadraticSymbolicExpression::from(T::zero());
+        for (var, mut coeff) in std::mem::take(&mut self.linear) {
+            if var == *variable {
+                let mut term = substitution.clone();
+                term *= &coeff;
+                to_add += term;
+            } else {
+                coeff.substitute(
+                    variable,
+                    substitution
+                        .try_to_known()
+                        .unwrap_or(&SymbolicExpression::from(T::zero())),
+                );
+                if !coeff.is_known_zero() {
+                    self.linear.insert(var, coeff);
+                }
```

Comment:
Same here, the variable being substituted should not appear in the coefficient, I think.